internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -0,0 +1,670 @@
+	pvc, mutateFn := r.newFilesystemPVC(instance)
+
+	if result, err := controllerutil.CreateOrUpdate(ctx, r.Client, pvc, mutateFn); err != nil {
+		conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
+			Type:    adiiov1alpha1.ConditionReconcileCompleted,
+			Status:  corev1.ConditionFalse,
+			Reason:  "StorageFailure",
+			Message: "Failed to ensure filesystem storage: " + err.Error(),
+		})
+		return err
+	} else if result != controllerutil.OperationResultNone {
+		r.Log.Info("Filesystem storage created")
+	}
+	return nil
+}
+
+func (r *AgentServiceConfigReconciler) ensureDatabaseStorage(ctx context.Context, instance *adiiov1alpha1.AgentServiceConfig) error {
+	pvc, mutateFn := r.newDatabasePVC(instance)
+
+	if result, err := controllerutil.CreateOrUpdate(ctx, r.Client, pvc, mutateFn); err != nil {
+		conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
+			Type:    adiiov1alpha1.ConditionReconcileCompleted,
+			Status:  corev1.ConditionFalse,
+			Reason:  "StorageFailure",
+			Message: "Failed to ensure database storage: " + err.Error(),
+		})
+		return err
+	} else if result != controllerutil.OperationResultNone {
+		r.Log.Info("Database storage created")
+	}
+	return nil
+}
+
+func (r *AgentServiceConfigReconciler) ensureAgentService(ctx context.Context, instance *adiiov1alpha1.AgentServiceConfig) error {
+	svc, mutateFn := r.newAgentService(instance)
+
+	if result, err := controllerutil.CreateOrUpdate(ctx, r.Client, svc, mutateFn); err != nil {
+		conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
+			Type:    adiiov1alpha1.ConditionReconcileCompleted,
+			Status:  corev1.ConditionFalse,
+			Reason:  "AgentServiceFailure",
+			Message: "Failed to ensure agent service: " + err.Error(),
+		})
+		return err
+	} else if result != controllerutil.OperationResultNone {
+		r.Log.Info("Agent service created")
+	}
+	return nil
+}
+
+func (r *AgentServiceConfigReconciler) ensureAgentRoute(ctx context.Context, instance *adiiov1alpha1.AgentServiceConfig) error {
+	route, mutateFn := r.newAgentRoute(instance)
+
+	if result, err := controllerutil.CreateOrUpdate(ctx, r.Client, route, mutateFn); err != nil {
+		conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
+			Type:    adiiov1alpha1.ConditionReconcileCompleted,
+			Status:  corev1.ConditionFalse,
+			Reason:  "AgentRouteFailure",
+			Message: "Failed to ensure agent route: " + err.Error(),
+		})
+		return err
+	} else if result != controllerutil.OperationResultNone {
+		r.Log.Info("Agent route created")
+	}
+	return nil
+}
+
+func (r *AgentServiceConfigReconciler) ensurePostgresSecret(ctx context.Context, instance *adiiov1alpha1.AgentServiceConfig) error {
+	// TODO(djzager): using controllerutil.CreateOrUpdate is convenient but we may
+	// want to consider simply creating the secret if we can't find instead of
+	// generating a secret every reconcile.
```

Comment:
Agree with this.
I think the secrets will need special handling so breaking the convention for these is probably okay.

If nothing else the empty mutate function might warrant a comment if that's what we're going with for the first pass.

Reply:
Added a comment to the mutateFn in `newPostgresSecret`.